app/src/main/java/org/mozilla/focus/shortcut/IconGenerator.kt

```diff
@@ -40,60 +33,36 @@ class IconGenerator {
-            canvas.drawBitmap(icon, drawX, drawY, Paint())
-
-            return shape
+            return drawCharacterOnBitmap(context, url, shape)
         }
 
         /**
          * Generates a launcher icon for versions of Android that support Adaptive Icons (Oreo+):
          * https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive.html
-         *
-         * A bitmap created from this drawable (e.g. Drawable.toBitmap) is expected to be used with
-         * [android.support.v4.graphics.drawable.IconCompat.createWithBitmap], not
-         * [android.support.v4.graphics.drawable.IconCompat.createWithAdaptiveBitmap]. Why? createWithAdaptiveBitmap
-         * will apply the platform's adaptive icon formatting to the given bitmap so it assumes you have a bitmap that
-         * hasn't been formatted already. However, when a bitmap is created with [AdaptiveIconDrawable], the formatting
-         * is applied so it'd be incorrect for us to use createWithAdaptiveBitmap, which would apply the formatting
-         * again.
-         *
-         * Unfortunately, we also can't just pass our legacy icon into createWithAdaptiveBitmap because the icon
-         * background has the corners removed and some device might display icons that have less of the corner missing
-         * than we do. In that case, we'd be unable to add the corners back and the icon wouldn't conform.
          */
-        @TargetApi(26)
-        private fun generateAdaptiveLauncherIcon(context: Context, url: String?): AdaptiveIconDrawable {
+        private fun generateAdaptiveLauncherIcon(context: Context, url: String?): Bitmap {
             val res = context.resources
-            val backgroundColor = ColorDrawable(ContextCompat.getColor(context, R.color.add_to_homescreen_icon_background))
-
             val adaptiveIconDimen = res.getDimensionPixelSize(R.dimen.adaptive_icon_drawable_dimen)
-            val foregroundLetter = BitmapDrawable(res, generateIcon(context, url, adaptiveIconDimen))
 
-            return AdaptiveIconDrawable(backgroundColor, foregroundLetter)
+            val bitmap = Bitmap.createBitmap(adaptiveIconDimen, adaptiveIconDimen, Bitmap.Config.ARGB_8888)
+            val canvas = Canvas(bitmap)
+
+            // Draw background to comply with Adaptive Icon guidelines
+            canvas.drawColor(res.getColor(R.color.add_to_homescreen_icon_background))
+
+            // Draw foreground to comply with Adaptive Icon guidelines
+            return drawCharacterOnBitmap(context, url, bitmap)
         }
 
-        /**
-         * Generate an icon for this website based on the URL.
-         */
-        private fun generateIcon(context: Context, url: String?, iconSize: Int): Bitmap {
-            val bitmap = Bitmap.createBitmap(iconSize, iconSize, Bitmap.Config.ARGB_8888)
+        private fun drawCharacterOnBitmap(context: Context, url: String?, bitmap: Bitmap): Bitmap {
             val canvas = Canvas(bitmap)
```

Comment:
I don't think anything needs to change here but I just want to note it: it's a little weird that we create a second Canvas (in the Adaptive icon case). But [the docs say](https://developer.android.com/reference/android/graphics/Canvas.html#Canvas(android.graphics.Bitmap)), "Construct a canvas with the specified bitmap to draw into." so it sounds like all the state is in the Bitmap, not the canvas, and this would be fine. It's a one time method so I don't think we care about the performance implications either.